src/cljam/io/bcf/writer.clj

```diff
@@ -98,104 +110,117 @@
-         type-byte (unchecked-byte (bit-or (bit-shift-left (min 15 total-len) 4) type-id))
-         bb (ByteBuffer/allocate (+ (* n-sample total-len (case type-id 1 1 2 2 3 4 5 4 7 1))
-                                    (if (<= 15 total-len) (case (int (value-type total-len)) 1 3 2 4 3 6) 1)))]
+     (encode-typed-value element-type [v] 1)))
+  (^bytes [element-type vs ^long n-sample]
+   (let [str? (or (= element-type :str) (= element-type :char))
+         vs (map (fn [v]
+                   (let [v (if (sequential? v) v [v])]
+                     (if str? (cstr/join \, (map #(or % ".") v)) v))) vs)
+         max-len (long (apply max 0 (map count vs)))
+         vs (map
+             (fn [v]
+               (let [l (max 1 (count v))]
+                 (if (< l max-len)
+                   (if str?
+                     (apply str v (repeat (- max-len l) (char 0)))
+                     (concat
+                      (or (seq v) [nil])
+                      (repeat (- max-len l) :eov)))
+                   v))) vs)
+         type-id (case element-type
+                   (:str :char) 7
+                   :float 5
+                   (long (apply max 1 (mapcat (partial map value-type) vs))))
+         type-byte (bit-or (bit-shift-left (min 15 max-len) 4) type-id)
+         len-bytes (when (<= 15 max-len)
+                     (encode-typed-value :int max-len))
+         n-bytes (+ (* n-sample max-len (case type-id 1 1 2 2 3 4 5 4 7 1))
+                    1 (if len-bytes (alength ^bytes len-bytes) 0))
+         bb (ByteBuffer/allocate n-bytes)]
      (.order bb ByteOrder/LITTLE_ENDIAN)
-     (.put bb type-byte)
-     (when (<= 15 total-len)
-       (.put bb ^bytes (encode-typed-value total-len)))
-     (doseq [x vs
-             b (if (or (sequential? x) (string? x)) x [x])]
+     (.put bb (unchecked-byte type-byte))
+     (when len-bytes
+       (.put bb ^bytes len-bytes))
+     (doseq [v vs
+             b v]
        (case type-id
          1 (.put bb (unchecked-byte (get int8-special-map b b)))
          2 (.putShort bb (unchecked-short (get int16-special-map b b)))
          3 (.putInt bb (unchecked-int (get int32-special-map b b)))
-         5 (.putFloat bb (float (get float32-special-map b b)))
-         7 (.put bb (byte b))))
+         5 (.putInt bb (unchecked-int (or (get float32-special-map b)
+                                          (Float/floatToRawIntBits b))))
+         7 (.put bb (byte (get {nil 0 :eov 0} b b)))))
      (.array bb))))
 
-(defn- encode-typed-kvs
-  "Encodes key-value pairs and returns as a byte-array."
-  [kvs]
-  (let [bas (->> kvs
-                 (mapcat
-                  (fn [[k v]]
-                    [(encode-typed-value k)
-                     (encode-typed-value v)])))
-        bb (ByteBuffer/allocate (reduce + (map alength bas)))]
-    (doseq [ba bas]
-      (.put bb ^bytes ba))
+(defn- merge-bytes
```

Comment:
It's just my personal opinion, but how about naming this function `concat-bytes`? The word `merge` implies (at least to me) the existence of a certain underlying merging strategy.